### DISCUSSION TITLE
Creating `pre_write_hooks`

### DISCUSSION THREAD

**Discussion by shruti222patel** [sqlalchemy/alembic#1114] [Usage Questions]
**Describe the use case**
I'd like to be able to mark certain columns of given tables as "frozen", so that when someone runs a migration dropping or changing those columns, the migration fails.

The project I work on is large and I want to ensure someone from another team doesn't accidentally delete columns my team depends on.

**Questions**
- Is a feature like "frozen" columns in the alembic backlog?
- I know alembic has `post_write_hooks`, I couldn't find docs for `pre_write_hooks`. Is that feature in the backlog? (If `pre_write_hooks` exist, I can write a custom one to solve this for myself.)

**Databases / Backends / Drivers targeted**

Postgres

**Comment by zzzeek** [comment]
so first off, for any approach, the way that you can "mark" a column with whatever you want is to use the ``.info`` dictionary.  So that looks like:

```py
Column("my_column", Integer, info={"frozen": True})
```

then the things we do after that will look for that .info in there.

Next, *when* you want to consume that is the issue.  the phrase "pre_write_hook" suggests you want to catch these columns during the autogenerate phase.  You can do that using [process_revision_directives](https://alembic.sqlalchemy.org/en/latest/api/autogenerate.html#customizing-revision-generation), so this sounds exactly like "pre write". 
  
however, when you say "when someone runs a migration, the migration fails", that's a different phase.  that means they wrote the migration and they are then running the script.      You can intercept actual "running" of things, like a `DropColumn` operation, using the `@compiles` SQLAlchemy hook.  However, if someone writes inside of a script `op.drop_column("table", "some_column")`, there's no indication there that that name is "frozen", because the model metadata is not involved.   if you intercepted operations at this level you'd need to somehow match those column names back to what's in your model to look for that "frozen" flag.  this could be done but is more complicated.

I think what you probably are looking for is the process_revision_directives hook above.

**Reply by shruti222patel** [reply]
Yes, you're correct @zzzeek  -- I actually want to to catch the columns as soon as possible. The autogenerate phase seems like it fits the bill. (I'm an alembic noob and so not familiar with the terms. 😅) 

My "frozen" columns will actually be generated by another service, and outputted as yaml or json. It seems like I should be able to create a "pre write" that reads that file and rejects the autogenerated commands if the columns exist in the yaml/json.

Would you please confirm I understood correctly?

**Reply by zzzeek** [reply]
you would implement a process_revision_directives hook that's connected within the env.py script.  this hook is called when someone uses autogenerate, and you'd want to traverse the structure that's generated and look for Column objects that have the "frozen" flag.  

now what *may* go wrong is that this info is not preserved when you get an op like AlterColumnOp, it should be there at least for CreateTableOp, so, take a look and if it's not there , I can figure something else out.
